backend/src/user/index.ts

```diff
@@ -18,9 +18,136 @@
+    if (res.rowCount === 0) {
+      return ErrorCode.USER_NOT_FOUND;
+    }
+    
+    let truePwHash = res.rows[0].pwhash;
+    let salt = new Buffer(res.rows[0].salt.data);
+
+    // hash the currently input password
+    let pwHash = await User.hashPassword(password, salt);
+    
+    if (pwHash === truePwHash) return ErrorCode.SUCCESS;
+    else {
+      // block execution for a second- this actually makes the system much, MUCH more secure
+      await timeout(1000);
+      return ErrorCode.PASSWORD_INCORRECT;
+    } 
+  }
+
+  // create a user from an object that has user-like properties (most likely an SQL row)
+  static fromRow(row: any): User {
+    return new User(row.user_id,
+                    row.username,
+                    row.email,
+                    row.karma,
+                    row.join_date,
+                    row.website,
+                    row.about,
+                    row.city,
+                    row.avatar,
+                    row.gender);
+  }
+
+  // load a user by its ID
+  static async loadById(user_id: Number): Promise<Nullable<User>> {
+    let res = await query("SELECT * FROM Users WHERE user_id=$1;", [user_id]);
+    if (res.rowCount === 0) return null;
+    else return User.fromRow(res.rows[0]);
+  }
+
+  // load a user by its username
+  static async loadByUsername(username: string): Promise<Nullable<User>> {
+    let res = await query("SELECT * FROM Users WHERE username=$1;", [username]);
+    if (res.rowCount === 0) return null;
+    else return User.fromRow(res.rows[0]);
+  }
+
+  // helper function- validate a user by its id or username
+  static async validateCredentials(user: Number | string, password: string): Promise<ErrorCode> {
+    let user_object: User;
+    if (user instanceof Number) user_object = await User.loadById(user);
+    else user_object = await User.loadByUsername(user);
+
+    return user_object.validate(password);
+  }
+
+  // add a new user to the database
+  // NOTE: number returned is an error code
+  static async createNewUser(username: string, 
+                             email: string, 
+                             password: string,
+                             return_user: boolean = false): Promise<ErrorCode | User> {
+    // check for user and email existence
+    let results = await Promise.all([checkUserExistence(username), checkEmailUsage(email)]);
+    if (results[0]) return ErrorCode.USER_EXISTS;
+    else if (results[1]) return ErrorCode.EMAIL_EXISTS;
+
+    // insert user into database
+    const addUserSql = "INSERT INTO Users (username, email, karma, join_date, status, avatar) " +
+                       "VALUES ($1, $2, 0, $3::timestamp, 0, '') RETURNING user_id;";
+    let res = await query(addUserSql, [username, email, getFormattedDate(new Date())]);
+    let user_id = res.rows[0].user_id;
+    
+    // generate salt and password hash
+    // NOTE TO FUTURE SELF: don't try to optimize this with Promise.all()
```

Comment:
I think
```suggestion
    // Password hashing is intentionally synchronous, do not use Promise.all()
```
is more clear